share/completions/p4.fish

```diff
@@ -0,0 +1,803 @@
+
+    # for now only looks that mode is set, later it will need to have a specific setting
+    echo 'threads='\t"sends files concurrently using N independent network connections"
+    echo 'batch='\t"specifies the number of files in a batch"
+    test "$mode"
+    or echo 'batchsize='\t"specifies the number of bytes in a batch"
+    echo 'min='\t"specifies the minimum number of files in a parallel sync"
+    test "$mode"
+    or echo 'minsize='\t"specifies the minimum number of bytes in a parallel sync"
+end
+
+#########################################################
+
+function __fish_print_p4_submit_options
+    echo 'submitunchanged'\t"Submit all open files (default behavior)"
+    echo 'submitunchanged+reopen'\t"Submit all open files + reopen to default changelist"
+    echo 'revertunchanged'\t"Revert unmodified files, submit all the rest"
+    echo 'revertunchanged+reopen'\t"Revert unmodified files, submit all the rest + reopen them in default changelist"
+    echo 'leaveunchanged'\t"Move unchanged files to default changelist, submit all the rest"
+    echo 'leaveunchanged+reopen'\t"Submit only modified files + reopen all (modified and unmodified) in default changelist"
+end
+
+#########################################################
+
+function __fish_print_p4_noretransfer_options
+    echo 1\t"server avoids re-transferring files that have already been archived after a failed submit operation"
+    echo 0\t"server re-transfers all files after a failed submit operation"
+end
+
+#########################################################
+
+function __fish_print_p4_integrate_output_options
+    echo b\t"outputs the base revision for the merge (if any)"
+    echo r\t"outputs the resolves that are being scheduled"
+end
+
+#########################################################
+
+function __fish_print_p4_integrate_resolve_options
+    echo b\t"schedules a branch resolve instead of branching the target files automatically"
+    echo d\t"schedules a delete resolve instead of deleting the target files automatically"
+    echo s\t"skips cherry-picked revisions that have already been integrated"
+end
+
+
+#########################################################
+
+function __fish_p4_in_subcommand
+    for i in (commandline -opc)
+        if contains -- $i (__fish_print_p4_commands_list)
+            return 0
+        end
+    end
+    return 1
+end
+
+#########################################################
+
+# Negating __fish_p4_in_subcommand in complete -n doesn't seem to work,
+# also commandline -opc doesn't work if __fish_p4_in_subcommand is called
+# from withing this function.
+# I must be missing something for DRY code.
+function __fish_p4_not_in_subcommand
+    for i in (commandline -opc)
+        if contains -- $i (__fish_print_p4_commands_list)
+            return 1
+        end
+    end
+    return 0
+end
+
+#########################################################
+
+function __fish_p4_is_using_command
+    set -l cmd $argv[1]
```

Comment:
This function can be rewritten as `contains -- "$argv[1]" (commandline -opc)`.

(Though what would be needed to properly model what p4 expects would probably be much more complicated, so keeping it as a function should be okay, so you can later expand it)